### DISCUSSION TITLE
Graph regression

### DISCUSSION THREAD

**Discussion by mtzortzi** [pyg-team/pytorch_geometric#3525] [Q&A] [answered]
Hi,

Thank you for all the work and help that you offer to all the community.

I'm having a dataset of graphs, which looks like this: Data(edge_attr=[444, 1], edge_index=[2, 444], x=[251, 4], y=[2]).
The node attributes are 4 as I wrote, but some of them are the same for all the nodes. This is because I had some features characterizing the whole graph (graph data) and I thought to add it to each node attribute so as to use pytorch geometric.

My goal is to predict those 2 values in y, which characterize the whole graph by taking some X graphs and predicting the y for the next X+3 graphs (3 is an example). I have divided my pytorch_geometric graphs into train, validation, and test sets and I have written some code using graph transformers.
 
But I'm a bit confused about how to approach the problem as long as the dataset is concerned. I suppose it's a regression problem and that the loss function I should use is MSE or MAE loss.
One of my ideas is to create masked values for y (for the next graphs of which the prediction I want), but in this case, what should I do with the edge_attr, edge_index, and x values of these graphs and their training part?

Could you please enlighten me with which approach I should tackle this problem?

Thank you for your time,
Marianna

**Comment by rusty1s** [comment, marked as answer]
Interesting problem. You could try to implement a dataset wrapper class, that returns 3 graphs and the labels of the fourth:
```python
class GroupDataset(torch.utils.data.Dataset):
    def __init__(self, dataset, window=3):
        super().__init__()
        self.dataset = dataset
        self.window = window
        
    def __len__(self):
        return len(self.dataset) - self.window
        
    def __getitem__(self, idx):
        data_list = [self.dataset[idx + offset] for offset in range(self.window)
        y = self.dataset[idx + self.window].y
        return data_list, y
```

**Reply by mtzortzi** [reply]
Thank you so much for your help! Indeed a dataset wrapper implementation worked. I used your structure and I wrote a class like the below one:

```python
class GroupDataset(torch.utils.data.Dataset):
    def __init__(self, dataset, window=3):
        super().__init__()
        self.dataset = dataset
        self.window = window #the number of trainable graphs
        
    def __len__(self):
        if self.dataset.num_graphs != self.window:
            return self.dataset.num_graphs - self.window
        else: 
            return self.dataset.num_graphs - (self.window-1)
        
    def __getitem__(self, idx):
        idx = idx - idx
        if self.dataset.num_graphs != self.window:
            data_list = [self.dataset[idx + offset] for offset in range(self.window)]
            y_list = [self.dataset[idx + self.window + restg].y for restg in range(self.dataset.num_graphs - self.window)]
        else:
            data_list = [self.dataset[idx+offset] for offset in range(self.window-1)]
            y_list = [self.dataset[idx + self.window + restg].y for restg in range(self.dataset.num_graphs - self.window -1)]
        return data_list, y_list
```

I made idx = 0 because I call this function from inside a PyTorch geometric dataloader so I want to create my dataset as X trainable graphs and Y.y predictable values for each batch of graphs. Also, I create a y_list for the values of y because I maybe want to predict y values for not 1 but more graphs.

Do you find this approach a good one?

I would also like to ask you about the last layer of my model. Should I have a readout layer to average all the node features of a graph?
I'm also thinking if a relu activation function would be enough for this task?

Thank you once again

**Reply by rusty1s** [reply]
I'm not sure I understand the line `idx = idx - idx` since this will make your `idx` to be always zero.
In case your targets correspond to graph-level labels, you will need a global pooling layer. It's probably best to apply global pooling for each graph separately, and then concatenate the output features of each graph into a sequence of graph-level features for further processing.

**Reply by mtzortzi** [reply]
Below I attach a part of my train_per_batch() function, in which I create the batch graphs from the PyTorch geometric dataloader. For every batchgraph I call the class GroupDataset() from above. So, yes maybe the line `idx = idx - idx` is a little unorthodox, but it's a way to iterate over each batch from inside the train_per_batch() function (so I keep the i), and also keep the right indexing for the graphs I want to return from inside the GroupDataset(). I hope now it is more clear.


``` python
    def train_per_batch(self, epoch, model, optimizer, train_pg_graphs, device):
        args = hubparser.hubparser()
        trainloader = DataLoader(train_pg_graphs, batch_size=args.batch_size, shuffle=True)
        epoch_loss = 0
        epoch_train_mae = 0
        model.train()
        for i, batchgraphs in enumerate(trainloader):
            start_time_training = time.time()
            batchgraphs = batchgraphs.to(device)
            print('========================')
            print(f'Step {i+1}:')
            dataset = GroupDataset(batchgraphs)
            print(f'Number of graphs in the current batch: {batchgraphs.num_graphs}')
            print(f'Number of graphs of which the y I want to predict: {len(dataset)}')
            traininputs, targets = dataset[i]
            optimizer.zero_grad() # Reset gradients
            
            # Passing the node features and the edge info
            batch_scores = model.forward(traininputs) 
            # Calculating the loss and the gradients
            loss_train = model.loss(batch_scores, targets)
            loss_train.backward()
            optimizer.step()
            epoch_loss += loss_train.detach().item()
            epoch_train_mae += MAE(batch_scores, targets)            
        epoch_loss /= (i+1)
        epoch_train_mae /= (i+1)

```

As long as the last layer is concerned, when you say to apply global pooling for each graph separately, do you mean for each mini-batch graph loaded from the dataloader?

**Reply by rusty1s** [reply]
I see. I'm not sure why you initialize `GroupDataset` in each batch once again. It should be possible to instantiate it only once, and then apply the conventional PyTorch training loop:
```python
dataset = GroupDataset(dataset, ...)
loader = torch_geometric.loader.DataLoader(dataset, batch_size=32)

for data_list, label in loader:
    # `data_list`: A list of batched graphs for every timestamp
    out = model(data_list)
    ...
```
Yes, I think it is beneficial to apply global pooling for every individual timestamp and concatenate the results afterwards.

**Reply by mtzortzi** [reply]
Hello,

Well, I thought I couldn't do it with pytorch_geometric, that's why I was doing it per batch. But now I instantiated it once as you told me.

I wanted to ask you about the  `out = model(data_list)`.  data_list is a list of DataBatch objects, which is not accessible, I mean, for example, I can not write data_list.x to extract the node_features. So, If my understanding is correct I can not insert to my model the data_list as it is. I have to insert it by iterating over it and getting the DataBatch objects inside my model. 

I would be really grateful if you could answer my question.

**Reply by rusty1s** [reply]
The `data_list` holds a `Batch` object for each timestamp, and you would need to iterate over it to compute output embeddings using either a shared or non-shared GNN, e.g.:
```python
out = []
for data in data_list:
    x = conv(data.x, data.edge_index).relu()
    x = conv(data.x, data.edge_index)
    x = global_mean_pool(x, data.batch)
    out.append(x)
out = torch.cat(out, dim=-1)
```

**Reply by mtzortzi** [reply]
Hello again,



About the iteration of data_list, this was my way of thinking too, I implemented the iteration inside my train function. 

``` python
    for step, (data_list, targets) in enumerate(trainloader):
        # `data_list`: A list of batched graphs for every timestamp
        batch_score_list = []
       for batch_graphs in data_list:
            batch_scores = model(batch_graphs) # load each DataBatch object for each list
```

I have 2 questions.

1. The first one concerns this implementation so far, with the DatasetWrapper as it is (taking 3 graphs and outputting the y label for the 4th one). 
I'm passing my batchgraphs in my model and after calculating the node_attr embeddings and edge_attr embeddings I use the TransformerConv from PyG docs. How could I insert the data.y from the graphs of the window (3 graphs) inside my TransformerConv? I suppose it's necessary if I have to predict the y of the next graph. Is my understanding correct? 

``` python
       ...
        self.layers = nn.ModuleList([TransformerConv(hidden_dim, hidden_dim, heads=heads, dropout=0.3, edge_dim=hidden_dim, norm=self.layernorm, beta=False)])
           for _ in range(1, num_layers):
            self.layers.append(TransformerConv(hidden_dim, out_dim, heads, dropout=0.3, edge_dim=hidden_dim, norm=self.layernorm, beta=False))
        ...

def forward(self, batch_graphs):
     ...
        for i, layer in enumerate(self.layers):
            h = F.dropout(h, p=0.5, training=self.training) #p=self.dropout
            h = layer(h, edge_index, e)

        if self.readout == "mean":
            global_x = global_mean_pool(h, batch_graphs.batch)[batch_graphs.batch]
            hg = torch.cat((x, global_x), dim=1)

```

2. The second question that I have is with different initial assumption. Imagine you have the graphs as it is e.g. Data(x=[823, 4], edge_index=[2, 1344], edge_attr=[1344, 1], y=[1, 2]). If I want to predict for every graph the data.y, what approach should I follow? 

P.S. the data.y in both cases 1 and 2 is a tensor [1, *] where * corresponds to 2 values, [timevalue, anotheroutputvalue]. That's why I can't go with a classification approach.

I'm sorry for the volume of questions in advance.

Your help to me is important and I really appreciate it.
Marianna

**Reply by rusty1s** [reply]
1. As mentioned earlier, a dataset example now holds a tuple containing a list of graphs and a ground-truth label of the 4th graph. When iterating, you will gain access to the ground-truth label as well:
```python
for data_list, y in loader:
    ...
```
2. This is similar to a classical graph-level prediction task, in which you stack a bunch of GNN layers, apply a global readout layer and train against the label of the current graph. It is similar to the first approach, except that you do not need to "merge" the embeddings coming from different graphs to make predictions.

I hope this is helpful :)

**Reply by mtzortzi** [reply]
Thank you so much for all the instructions. 

I think it is clearer to me now. So for both the implementations, I need the global readout layer to train against the labels of the batch graphs (1st approach), or the label of the current graph (2nd approach). But in the 1st approach, I need the global merging.

I'm training now the 1st approach, but while the TransformerConv part of the code is implemented this bug arises. (The TransformerConv is the one from PyTorch Geometric docs, I just took the source code for further investigation).

```
   File "/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/GraphTransformerModel.py", line 108, in forward
    h = self.conv1(h, edge_index, e)
  File "/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/layers/Graph_transformer_layer.py", line 143, in forward
    out += x_r
RuntimeError: The size of tensor a (69376) must match the size of tensor b (4) at non-singleton dimension 1
```

My node embeddings have a shape of torch.Size([17344, 4, 64])  and 
my edge embeddings a shape of torch.Size([28320, 1, 64]). 
When are passed into the TransformerConv with heads=8 and output_channels=hidden_dim=64 query is of shape torch.Size([69376, 8, 64]).  
The line 143 where the error hits is supposed to take the x_r = self.lin_skip(x[1]) and eventually add it to out += x_r. 

I'm missing something with the dimensions. Would you please have a look?

**Reply by rusty1s** [reply]
Node and edge features should be two-dimensional if passed to a GNN, which likely explains the error. Any reason they are three-dimensional in your case?

**Reply by mtzortzi** [reply]
Well, my initial tensors are e.g.: x of shape torch.Size([17344, 4]) and edge_attr torch.Size([28320,1]). 

For the embeddings I use 

``` python
self.embedding_h = torch.nn.Embedding(n_input_dim+1, hidden_dim)
self.embedding_e = torch.nn.Embedding(e_input_dim+1, hidden_dim)
```
where n_input_dim is the size of the dictionary of embeddings (number of nodes or edges)
and hidden_dim is the size of each embedding vector.
So when I call the embedding from the forward function 

`h = self.embedding_h(x)`

to calculate the embeddings the resulting tensors for each of my input tensors are:
for node embeddings torch.Size([17344, 4, 64]) and 
for edge embeddings torch.Size([28320, 1, 64]).

These are the tensors that are passed into the TransformerConv.

**Reply by rusty1s** [reply]
I see. You should either reshape them to `torch.Size([17344, 256])` and `torch.Size([28320, 64])`, or sum them up in the second dimension.

**Reply by mtzortzi** [reply]
Thank you for your instructions. I had already tried the reshape with .view() and it doesn't work inside the Transformer Conv due to dimensional conflicts, but the summation in the second dimension works fine and I didn't think about it before. As I'm interested in inserting positional encoding and other types of encoding inside the TransformerConv layer, I would like to ask you whether this could be possible.

**Reply by rusty1s** [reply]
Well, this depends on how you define positional encodings in a graph :) There exists some work of that in the field, e.g., [here](https://arxiv.org/abs/2010.02863), but this isn't yet integrated into PyG.

**Reply by mtzortzi** [reply]
Yes, you're right. I'll have a look and I'll come back to you. Thank you again for your help.
